### PR TITLE
Share GetConstructor intrinsic a lot of small fixes

### PR DESCRIPTION
**Pull request by vitek-karas**
The main change is sharing of the GetConstructor intrinsic and related tests.
To make those tests work this also implements Type.EmptyTypes and String.Empty and Array.Empty intrinsics.
These are all treated as special cases:
* Type.EmptyTypes and String.Empty are actually field accesses, so we don't have infra to share these yet, no need to add one for now
* Array.Empty is a generic method where we need to know the generic instantiation. The current sharing doesn't propagate this information
  so this forced the implementation to NOT be shared but instead done as a fallback which the shared one doesn't handle.

The rest of the changes are all about fixing how the shared intrinsic handling propagates return value from intrinsics.
Lot of small subtle bugs around null or emtpy inputs and forgetting to add return value and so on. The main changes are in GetMethod and GetNestedType intrinsics which had lot of issues.

Added tests for all of the corner cases.
NestedTypesUsedViaReflection test has been reworked as previously it didn't work correctly. Lot of the [Kept] attributes were fulfilled by some of the negative tests and thus the positive tests didn't have coverage (for example calling GetNestedType(unknownName) will mark all nested types on the type, so a subsequent call to GetNestedType("knownname") will not have an observable effect).

### CODE REVIEW

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 1]
```diff
@@ -302,14 +302,15 @@ public override bool HandleCall (MethodBody callingMethodBody, MethodReference c
...
+			case IntrinsicId.Type_get_BaseType:
+			case IntrinsicId.Type_GetConstructor: {
 					var instanceValue = MultiValueLattice.Top;
 					IReadOnlyList<MultiValue> parameterValues = methodParams;
 					if (calledMethodDefinition.HasImplicitThis ()) {
 						instanceValue = methodParams[0];
 						parameterValues = parameterValues.Skip (1).ToImmutableList ();
 					}
-					return handleCallAction.Invoke (calledMethodDefinition, instanceValue, parameterValues, out methodReturnValue);
+					return handleCallAction.Invoke (calledMethodDefinition, instanceValue, parameterValues, out methodReturnValue, out _);
```
Comment (sbomer) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:313]: Would it make sense to always call `handleCallAction.Invoke`, and then handle any non-shared intrinsics, to avoid duplicating the cases above?

File: src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs [hunk 2]
```diff
@@ -665,6 +706,51 @@ GenericParameterValue genericParam
...
+						1 => (argumentValues[0].AsSingleValue () as ArrayValue)?.Size.AsConstInt (),
+						2 => (argumentValues[1].AsSingleValue () as ArrayValue)?.Size.AsConstInt (),
+						4 => (argumentValues[2].AsSingleValue () as ArrayValue)?.Size.AsConstInt (),
+						5 => (argumentValues[3].AsSingleValue () as ArrayValue)?.Size.AsConstInt (),
+						_ => null,
+					};
+
+					// Go over all types we've seen
+					foreach (var value in instanceValue) {
+						if (value is SystemTypeValue systemTypeValue && !BindingFlagsAreUnsupported (bindingFlags)) {
```
Comment (sbomer) [inline src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:733]: Does this also need handling for null/empty inputs?

File: src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs [hunk 3]
```diff
@@ -357,6 +363,11 @@ GenericParameterValue genericParam
 			// GetMethod (string, int, BindingFlags, Binder?, CallingConventions, Type[], ParameterModifier[]?)
 			//
 			case IntrinsicId.Type_GetMethod: {
+					if (instanceValue.IsEmpty () || argumentValues[0].IsEmpty ()) {
```
Comment (sbomer) [inline src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:366]: Thanks a lot for the null/empty handling cleanup in this PR!

**Review by sbomer — APPROVED** [review]

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs (hunk 1 above)
Comment (vitek-karas) [reply src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:313]: It does make sense - but I didn't want to do it in this change. That should happen when we have shared all of the intrinsics and those which can't be shared should all switch to the same behavior as the Array.Empty does in this change.
Similarly to MakeGenericMethod - it's not finished yet, and it doesn't handle empty input well yet - again, once all of that is shared we will fix it up.

The change was already getting out of hand as is, I didn't want to make it even bigger.

File: src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs (hunk 2 above)
Comment (vitek-karas) [reply src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:733]: Not really - we could add it, but it's not necessary. We need special handling only for intrinsics which set return value, those which don't can have empty foreach loops, it doesn't cause a problem.

Not sure what's better - be consistent with unnecessary lines of code, or keep it shorter.

Comment (sbomer) [reply src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:733]: Oh right, the difference is that it doesn't have a return value. My preference is consistency, but either way is fine.
